### DISCUSSION TITLE
Mask image. Detect circles and mask them out.

### DISCUSSION THREAD

**Discussion by TNNSLAB** [ImageMagick/ImageMagick#5188] [Help] [answered]
I need to mask everything out and keep the skate deck only. 

Also, those top two circles I would need to detect them and mask them out too because they have some pins. 

Any ideas ? 

![skate-165](https://user-images.githubusercontent.com/46636486/171819655-602e474d-9706-4fc8-9443-a963106072ce.jpg)

<img width="1366" alt="Screenshot 2022-06-03 at 10 42 31" src="https://user-images.githubusercontent.com/46636486/171820489-5d87c09d-5b6a-4dda-ac61-3d75f880e2ee.png">

**Comment by fmw42** [comment]
What version of Imagemagick and what OS?

**Comment by TNNSLAB** [comment]
Hi Fred, 

Version: ImageMagick 7.1.0-4 Q16 x86_64 2021-07-18
Mac OS X 10.14.6 Mojave

**Reply by fmw42** [reply]
I am not sure I understand the question. Do you want the circle removed or to find the circles? Or something else.

**Reply by TNNSLAB** [reply]
Ok, let me try to explain again. 
So first of all, i would need to detect that board an delete the backround around it so that is on a png alpha. 
Second, if possible, I would need to detect the very first top 2 circles and fill them white because right now there's some pins in them.

**Comment by snibgo** [comment]
Why not do the job with Gimp?

**Reply by TNNSLAB** [reply]
i want to automate this process…

**Comment by snibgo** [comment]
Okay, I suppose you have many images you want to apply this to. What are the common features that define the "skate deck" on all the images? Perhaps it is defined by:

1. Find the largest nearly-white area.

2. The skate deck is the collection of dark pixels within that white area, and all the pixels within that dark area.

If that is a reasonable definition, one approach would be: use connected components to find the bounding box of the large white area. Crop to that box, and trim the white away. That gives the bounding box of the skate deck.

Add a white boundary, then flood-fill from top-left, changing white to an unused colour (eg red). Then all he pixels that are not red are pixels within the skate deck.

Detecting the small circles needs similar analysis. What defines them? What defines those circles, and not the other pair of circles?

**Reply by TNNSLAB** [reply]
A common features that defines skate deck would be the option **nr 2** you described : 
_"The skate deck is the collection of dark pixels within that white area, and all the pixels within that dark area."_

I'm sure your approach would work. I'll have to figure out how am I doing this now. 

I found something like : 
`-define connected-components:area-threshold=200000`

Or you do it the other way ?

**Comment by fmw42** [comment, marked as answer]
This will get you started. I left some tmp files in there so you can see the steps.

```
bbox=$(magick skate.jpg \
-auto-threshold otsu -type bilevel \
-morphology open disk:3 \
-define connected-components:verbose=true \
-define connected-components:mean-color=true \
-define connected-components:keep-top=3 \
-connected-components 8 null: | grep "gray(255)" | awk '{print $2}')

echo $bbox

magick skate.jpg -crop $bbox +repage +write tmp1.png \
-shave 250x250 +write tmp2.png \
-fuzz 45% -trim +repage \
skate_trimmed.png

```

![skate_trimmed](https://user-images.githubusercontent.com/13458218/172652475-0a897b21-8c7f-408f-8716-a77393ddeecd.png)

**Reply by TNNSLAB** [reply]
Phenomenal ! As usual ! Thank you very much !

**Comment by fmw42** [comment]
Sorry, I was in a hurry before and did not have time to explain. According to the suggestions by snibgo:

```
I first threshold the input automatically
Next, I apply some morphology to fill out the black areas where they are weak
Then, I get the top 3 connected components (black, white, black) and filter to keep only the white or gray(255) and extract only the bounding box and put it in a variable.
Then, I take the input again and crop it using the bounding box, 
Then I shave the sides all around to get rid of excess black in the white area
Then I trim the white to the bounding box of the central black area.
Then save the result

```

```
bbox=$(magick skate.jpg \
-auto-threshold otsu -type bilevel \
-morphology open disk:3 \
-define connected-components:verbose=true \
-define connected-components:mean-color=true \
-define connected-components:keep-top=3 \
-connected-components 8 null: | grep "gray(255)" | awk '{print $2}')

echo $bbox

magick skate.jpg -crop $bbox +repage +write tmp1.png \
-shave 250x250 +write tmp2.png \
-fuzz 45% -trim +repage \
skate_trimmed.png

```
